cache/redis_test.go

```diff
@@ -0,0 +1,491 @@
+			args: args{
+				artifactID: "sha256:961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4",
+				blobIDs:    []string{"sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0"},
+			},
+			wantMissingArtifact: true,
+		},
+		{
+			name:       "missing blobs",
+			setupRedis: true,
+			args: args{
+				artifactID: "sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf",
+				blobIDs:    []string{"sha256:1b3ee35aacca9866b01dd96e870136266bde18006ac2f0d6eb706c798d1fa3c3"},
+			},
+			wantMissingArtifact: false,
+			wantMissingBlobIDs:  []string{"sha256:1b3ee35aacca9866b01dd96e870136266bde18006ac2f0d6eb706c798d1fa3c3"},
+		},
+		{
+			name:       "missing artifact with different schema version",
+			setupRedis: true,
+			args: args{
+				artifactID: "sha256:be4e4bea2c2e15b403bb321562e78ea84b501fb41497472e91ecb41504e8a27c",
+				blobIDs:    []string{"sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0"},
+			},
+			wantMissingArtifact: true,
+		},
+		{
+			name:       "missing blobs with different schema version",
+			setupRedis: true,
+			args: args{
+				artifactID: "sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf",
+				blobIDs:    []string{"sha256:174f5685490326fc0a1c0f5570b8663732189b327007e47ff13d2ca59673db02"},
+			},
+			wantMissingArtifact: false,
+			wantMissingBlobIDs:  []string{"sha256:174f5685490326fc0a1c0f5570b8663732189b327007e47ff13d2ca59673db02"},
+		},
+	}
+
+	// Set up Redis test server
+	s, err := miniredis.Run()
+	require.NoError(t, err)
+	defer s.Close()
+
+	s.Set("fanal::artifact::sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf", `{"SchemaVersion": 1}`)
+	s.Set("fanal::artifact::sha256:be4e4bea2c2e15b403bb321562e78ea84b501fb41497472e91ecb41504e8a27c", `{"SchemaVersion": 2}`)
+	s.Set("fanal::blob::sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0", `{"SchemaVersion": 1}`)
+	s.Set("fanal::blob::sha256:174f5685490326fc0a1c0f5570b8663732189b327007e47ff13d2ca59673db02", `{"SchemaVersion": 2}`)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			addr := s.Addr()
+			if !tt.setupRedis {
+				addr = "dummy:6379"
+			}
+
+			c := cache.NewRedisCache(&redis.Options{
+				Addr: addr,
+			})
+
+			missingArtifact, missingBlobIDs, err := c.MissingBlobs(tt.args.artifactID, tt.args.blobIDs)
+			if tt.wantErr != "" {
+				require.NotNil(t, err)
+				assert.Contains(t, err.Error(), tt.wantErr)
+				return
+			} else {
+				assert.NoError(t, err)
+			}
+
+			assert.Equal(t, tt.wantMissingArtifact, missingArtifact)
+			assert.Equal(t, tt.wantMissingBlobIDs, missingBlobIDs)
+		})
+	}
+}
+
+func TestRedisCache_Close(t *testing.T) {
+	// Set up Redis test server
+	s, err := miniredis.Run()
+	require.NoError(t, err)
+	defer s.Close()
+
+	t.Run("close", func(t *testing.T) {
+		c := cache.NewRedisCache(&redis.Options{
+			Addr: s.Addr(),
+		})
+		require.NoError(t, c.Close())
```

Comment:
This one is actually testing that the `Close()` method did not return any error. Not sure if it's worth it, but it seems that miniredis allows you to check current connections so you could write:

```suggestion
		closeErr := c.Close()
		require.NoError(t, closeErr)
		time.Sleep(3 * time.Second) // give it some time
		assert.Equal(t, 0,s.CurrentConnectionCount(), "The client is disconnected")
```